File: java/L1000/LeetCode1035.java
package L1000;

public class LeetCode1035 {
    public int maxUncrossedLines(int[] nums1, int[] nums2) {
        final int M = nums1.length, N = nums2.length;
        int[][] dp = new int[M + 1][N + 1];
        for (int i = 1;i <= M;++i) {
            for (int j = 1;j <= N;++j) {
                if (nums1[i - 1] == nums2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[M][N];
    }
}

File: cpp/L1000/LeetCode1035.cc
#include <map>
#include <vector>
using std::vector;

class Solution {
public:
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        std::map<int, vector<int>> mp;
        int N2 = nums2.size();
        for (int i = N2 - 1;i >= 0;--i) {
            mp[nums2[i]].push_back(i);
        }
        vector<int> counts(N2, 0);
        int ans{0};
        for (int num : nums1) {
            for (int ix : mp[num]) {
                counts[ix] = 1;
                for (int iy = ix - 1;iy >= 0;--iy) {
                    counts[ix] = std::max(counts[ix], counts[iy] + 1);
                }
                ans = std::max(ans, counts[ix]);
            }
        }
        return ans;
    }
};